js/modules/loki_p2p_api.js

```diff
@@ -18,68 +19,59 @@ class LokiP2pAPI extends EventEmitter {
-        port,
-        timerDuration,
-        pingTimer: null,
-        isOnline: false,
-      };
-      if (isPing) {
-        this.setContactOnline(pubKey);
-        return;
-      }
-      // Try ping
-      this.pingContact(pubKey);
-      return;
-    }
-
-    // We already had this contact's details
-    const baseDetails = { ...this.contactP2pDetails[pubKey] };
-
-    if (isPing) {
-      // Received a ping
-      // Update details in case they are new and mark online
-      this.contactP2pDetails[pubKey].address = address;
-      this.contactP2pDetails[pubKey].port = port;
+    // Get the current contact details
+    // This will be empty if we don't have them
+    const baseDetails = { ...(this.contactP2pDetails[pubKey] || {}) };
+
+    // Always set the new contact details
+    this.contactP2pDetails[pubKey] = {
+      address,
+      port,
+      timerDuration,
+      pingTimer: null,
+      isOnline: false,
+    };
+
+    const contactExists = !isEmpty(baseDetails);
+    const { isOnline } = baseDetails;
+    const detailsChanged =
+      baseDetails.address !== address || baseDetails.port !== port;
+
+    // If we had the contact details
+    // And we got a P2P message
+    // And the contact was online
+    // And the new details that we got matched the old
+    // Then we don't need to bother pinging
+    if (contactExists && isP2PMessage && isOnline && !detailsChanged) {
+      // We also need to set the current contact details to show online
+      //  because they get reset to `false` above
       this.setContactOnline(pubKey);
       return;
     }
 
-    // Received a storage broadcast message
-    if (
-      baseDetails.isOnline ||
-      baseDetails.address !== address ||
-      baseDetails.port !== port
-    ) {
-      // Had the contact marked as online and details we had were the same
-      this.pingContact(pubKey);
-      return;
-    }
-
-    // Had the contact marked as offline or got new details
-    this.contactP2pDetails[pubKey].address = address;
-    this.contactP2pDetails[pubKey].port = port;
-    this.setContactOffline(pubKey);
+    /*
+      Ping the contact.
+      This happens in the following scenarios:
+        1. We didn't have the contact, we need to ping them to let them know our details.
+        2. isP2PMessage = false, so we assume the contact doesn't have our details.
+        3. We had the contact marked as offline,
+            we need to make sure that we can reach their server.
+        4. The other contact details have changed,
+            we need to make sure that we can reach their new server.
+    */
     this.pingContact(pubKey);
   }
 
   getContactP2pDetails(pubKey) {
-    return this.contactP2pDetails[pubKey] || null;
-  }
-
-  isContactOnline(pubKey) {
-    const contactDetails = this.contactP2pDetails[pubKey];
-    return !!(contactDetails && contactDetails.isOnline);
+    if (!this.contactP2pDetails[pubKey]) return null;
+    return { ...this.contactP2pDetails[pubKey] };
```

Comment:
Was this returning a reference instead of a copy?
If so, good catch.

Reply:
Ya, don't think it was affecting the actual functionality but it was picked up by the tests